### PR TITLE
opt: allow zero-length payload in LengthFieldBasedFrameCodec

### PR DESCRIPTION
**Pull request by firedtoad**
allow zero length playload

---
name: Pull request
about: Propose changes to the code
title: ''
labels: ''
assignees: ''
---

## 4. Checklist

- [ ] I have squashed all insignificant commits.
- [ ] I have commented my code for explaining package types, values, functions, and non-obvious lines.
- [ ] I have written unit tests and verified that all tests passes (if needed).
- [ ] I have documented feature info on the README (only when this PR is adding a new feature).
- [ ] (optional) I am willing to help maintain this change if there are issues with it later.

### CODE REVIEW

**Comment by panjf2000** [comment]
What's the point of allowing zero `readN` call?

**Comment by firedtoad** [comment]
There is zero length playload packat sometime, like heartbeat for keeping
connection alive.


Andy Pan <notifications@github.com> 于2020年7月7日周二 下午1:24写道：

File: codec.go [hunk 1]
```diff
@@ -222,8 +222,11 @@ func (cc *LengthFieldBasedFrameCodec) Encode(c Conn, buf []byte) (out []byte, er
 type innerBuffer []byte
 
 func (in *innerBuffer) readN(n int) (buf []byte, err error) {
-	if n <= 0 {
-		return nil, errors.New("zero or negative length is invalid")
+	if n==0 {
```
Comment (panjf2000) [inline codec.go:225]: It seems that this line wasn't formatted, it should be `n == 0` instead.

Comment (panjf2000) [reply codec.go:225]: Run `gofmt -l -w -s codec.go`.

**Comment by panjf2000** [comment]
Besides, have you tested this PR with the case of zero-length payload? did it work?

**Comment by firedtoad** [comment]
Yes of course, I found it is a bug in my working project.

**Comment by panjf2000** [comment]
Why did you close this PR?

**Comment by panjf2000** [comment]
Sorry, I forgot that the unit tests of `Decode()` for codec are still unfinished, never mind, I will write those unit tests.

File: codec.go [hunk 2]
```diff
@@ -222,8 +222,11 @@ func (cc *LengthFieldBasedFrameCodec) Encode(c Conn, buf []byte) (out []byte, er
 type innerBuffer []byte
 
 func (in *innerBuffer) readN(n int) (buf []byte, err error) {
-	if n <= 0 {
-		return nil, errors.New("zero or negative length is invalid")
+	if n == 0 {
+		return []byte{}, nil
```
Comment (panjf2000) [inline codec.go:226]: It can be `return nil, nil` directly, no need to initialize an empty []byte, standard Go functions accept nil buffers, they can be used just like []byte{}.

**Comment by panjf2000** [comment]
I've added unit tests for `Decode(Conn)`, you can add your unit test for zero-length payload case now, you can take a reference from https://github.com/panjf2000/gnet/blob/master/codec_test.go#L73-L82.
